File: 174.dungeon-game.py
#
# @lc app=leetcode id=174 lang=python3
#
# [174] Dungeon Game
#

# @lc code=start
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        height, width = len(dungeon), len(dungeon[0])
        dp = [1 << 20]*(width+1)
        dp[width-1] = 1
        # backward induction
        for i in range(height-1, -1, -1):
            for j in range(width-1, -1, -1):
                # decide the minimum health power if knight "start at here"
                dp[j] = min(dp[j], dp[j + 1]) - dungeon[i][j]
                dp[j] = max(dp[j], 1)
        return dp[0]
# @lc code=end

File: java/174.dungeon-game.java
import java.util.Arrays;

/*
 * @lc app=leetcode id=174 lang=java
 *
 * [174] Dungeon Game
 */

// @lc code=start
class Solution {
    public int calculateMinimumHP(int[][] dungeon) {
        int height = dungeon.length, width = dungeon[0].length;
        int[] dp = new int[width + 1];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[width - 1] = 1;
        // backward induction
        for (int i = height - 1; i != -1; --i) {
            for (int j = width - 1; j != -1; --j) {
                // decide the minimum health power if knight "start at here"
                dp[j] = Math.min(dp[j], dp[j + 1]) - dungeon[i][j];
                dp[j] = Math.max(dp[j], 1);
            }
        }
        return dp[0];
    }
}
// @lc code=end
